### ISSUE TITLE
TestClient DELETE method doesn't support any payload

### ISSUE THREAD

**Issue by andybbruno** [tiangolo/fastapi#5649]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
--- APP ---

from fastapi import FastAPI

app = FastAPI()

@app.delete("/{user_id}")
async def _delete_user(user_id: int):
    return {}



--- TEST ---

def test_delete_user(client: TestClient):
    user_id = 123
    response = client.delete(f"/{user_id}", json={"validation_code": 789})
    assert response.status_code == 200
```

### Description

```
TypeError: TestClient.delete() got an unexpected keyword argument 'json'
```

After updating my application to FastAPI 0.87.0 I found out that the new `TestClient` does not support anymore any kind of payload in the `DELETE` method. Although the use of a body within DELETE requests is highly discouraged, it is not forbidden. Hence, we should find a way to allow a full retro-compatibility. 

An idea would be to wrap the `httpx.delete` method used within the `TestClient` class in order to re-enable this feature. For this to happen, the solution described [here](https://www.python-httpx.org/compatibility/#request-body-on-http-methods) can be reused.
  

### Operating System

macOS

### FastAPI Version

0.87.0

### Python Version

3.10

**Comment by mbroton** [comment]
[FastAPI's release 0.87.0](https://github.com/tiangolo/fastapi/releases/tag/0.87.0) upgrades Starlette to [0.21.0](https://github.com/encode/starlette/releases/tag/0.21.0) which replaces `requests` with `httpx`.
`fastapi.testclient.TestClient` is only a shortcut for `starlette.testclient.TestClient`, thus, IMO, you should open an issue in Starlette's issues. You'll also probably receive an answer faster there.

**Comment by yinziyan1206** [comment]
[https://github.com/encode/httpx/discussions/1587](url)

**Comment by andybbruno** [comment]
> [FastAPI's release 0.87.0](https://github.com/tiangolo/fastapi/releases/tag/0.87.0) upgrades Starlette to [0.21.0](https://github.com/encode/starlette/releases/tag/0.21.0) which replaces `requests` with `httpx`.
> 
> `fastapi.testclient.TestClient` is only a shortcut for `starlette.testclient.TestClient`, thus, IMO, you should open an issue in Starlette's issues. You'll also probably receive an answer faster there.

Thanks @mbroton for the suggestion. I'm quite sure that if I'll open an issue on Starlette repo they'll tell me to go back to FastAPI repo 😅

**Comment by andybbruno** [comment]
> [https://github.com/encode/httpx/discussions/1587](url)

As I said, using payloads is highly discouraged but not forbidden!

**Comment by yinziyan1206** [comment]
> > [https://github.com/encode/httpx/discussions/1587](url)
> 
> As I already said, using payloads is highly discouraged but not forbidden! Furthermore, before this update everything worked fine
Yes, but `httpx` discourages it:(.  maybe you can use `request` instead of `delete`.

**Comment by andybbruno** [comment]
> Yes, but `httpx` discourages it:(

That's why I opened this issue on the FastAPI repo ;)

**Comment by jbakermk** [comment]
I just came across exactly the same issue after updating my project dependencies. Someone has removed the json attribute from the TestClient.delete function. You can work around it as follows:

`client.request("DELETE", "http://service.endpoint", json={..})`

but it's not ideal.

**Comment by rani-pinchuk** [comment]
> > [https://github.com/encode/httpx/discussions/1587](url)
> 
> As I already said, using payloads is highly discouraged but not forbidden! Furthermore, before this update everything worked fine

I suspect that also the GET method does not support now payloads. I get the following in my tests now.
```
response = pytest.client.get("/myfunc", json={"some-key": "some-value"})
E       TypeError: get() got an unexpected keyword argument 'json'
```
It is (maybe discouraged but) allowed to use payloads with GET, and this was working with FastAPI. Still, since the introcution of `httpx` in FastAPI, our tests are broken.

As this is a major change of the API, I think that if this change is kept, it deserves a major change to the version of fastapi.

**Comment by chbndrhnns** [comment]
Doesn't 0.x.x imply that breaking changes could occur anytime?

**Comment by andybbruno** [comment]
> Doesn't 0.x.x imply that breaking changes could occur anytime?

Technically not.. 

![1589209257533](https://user-images.githubusercontent.com/28092511/202511133-26bc7f1b-82cb-4acd-a879-dc189c4978aa.png)

**Comment by Kludex** [comment]
> Doesn't 0.x.x imply that breaking changes could occur anytime?

It does. That's why it's documented to [pin the FastAPI version](https://fastapi.tiangolo.com/deployment/versions/?h=pin#pin-your-fastapi-version).

The semver is not applied until 1.0, so the image above doesn't count.

> I'm quite sure that if I open an issue on Starlette repo they'll tell me to go back to FastAPI repo sweat_smile

Given that I'm the most active contributor on Starlette, I can assure you I wouldn't. :)

If you want to send a body with a `DELETE` method, you can replace `client.delete(...)` by `client.request(...)`, as mentioned in the README of [bump-testclient](https://github.com/Kludex/bump-testclient):
![image](https://user-images.githubusercontent.com/7353520/202515799-edbd1e73-ac18-40b0-8c6a-eaed2e50df93.png)

**Comment by andybbruno** [comment]
Thanks for your help @Kludex!

> The semver is not applied until 1.0, so the image above doesn't count.

Sorry, I didn't know about it.  Anyways, [according to `semver`](https://semver.org/#how-do-i-know-when-to-release-100) FastAPI can be released with version 1.0.0

**Comment by rani-pinchuk** [comment]
The argument that this is zero version and therefore you can break the API at any time is correct. 

But it is not convincing. It looks like FastAPI **is** used in production. On FastAPI main page, we can see that it is already used by major player (Netflix, Uber...). I respect very much the modesty in being careful not to go too early to version 1+. But I think that waiting too long is a mistake.

In the roadmap of FastAPI, I cannot find the milestone to move up to 1.x.x., and until this discussion, I must admit that I did not realize that it is still dangerous to use FastAPI due to the risk that the API will be changed without warnings. 

Is there a good reason today to keep the major version as zero?

**Comment by tiangolo** [comment]
Thanks for the feedback @rani-pinchuk 

What is the risk that would not happen in a version 1.x.x? And what do you mean with a warning? What would you expect and what makes it different if the string in the version number starts with 1.x.x?

I want to understand what are your expectations for a version 1.x.x and how having that as the string number would be technically better right away rather than a version 0.x.x.

Also, the information of how to properly and carefully pin and upgrade versions, in particular for prod deployments, is here: https://fastapi.tiangolo.com/deployment/versions/

**Comment by rani-pinchuk** [comment]
I am not totally sure. You can always say that the change to the API that was made above was to remove a behavior that was anyway not totally supported (payload on GET and DELETE methods). 

But if it was done on 1.x.x or later, I personally would expect that this would result in a major version change (as API was changed, and this has the potential to break things). This in turn would save me some time trying to understand how my tests were suddenly broken. Again, you can argue that this is not a real breaking change, as the behavior was not really supported. 

The page about the versions indeed explain well that each version could potentially have breaking changes. But it also states that "FastAPI is already being used in production in many applications and systems". This means that it should be with a major version bigger than zero. 

I understand that features are still being added, but then, when there is a breaking change, you can just upgrade the major version number.

**Comment by StefanEmbed** [comment]
"So, yep, you can still send payloads with HTTP operations that don't really support it, but you have to be explicit that you are doing something that is kind of wrong. That's also why it's not part of HTTPX by default, to stop encouraging developers to use HTTP in a way that is kinda wrong and possibly broken in some environments that are more strict about the standard."
^
This is why we can't have nice things, bs standards and best practices that really just end up complicating things needlessly.
Why should I not be able to send a list of ids that I want removed? Why do I HAVE to delete each individual object by itself and spam the internet with hundreds of delete requests. Makes no sense.
